### PR TITLE
Hact v2

### PR DESCRIPTION
**Pull request by domdinicola**

### CODE REVIEW

File: EquiTrack/hact/admin.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from django.contrib import admin
```
Comment (reinbach) [inline EquiTrack/hact/admin.py]: Don't forget to add the `__future__` imports on new files

File: EquiTrack/hact/views.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            if query_params.get("format") == 'csv':
+                response['Content-Disposition'] = "attachment;filename={}.csv".format(self.filename)
+        return response
+
+
+class GlobalHactView(views.APIView):
+    permission_classes = []
+
+    def get(self, request, *args, **kwargs):
+        return Response(
```
Comment (reinbach) [inline EquiTrack/hact/views.py]: How come the values are hard coded here?

File: EquiTrack/partners/models.py [hunk 3]
```diff
@@ -378,6 +416,16 @@ class PartnerOrganization(AdminURLMixin, models.Model):
 
     tracker = FieldTracker()
 
+    @cached_property
+    def expiring_assessment_flag(self):
```
Comment (reinbach) [inline EquiTrack/partners/models.py]: There is no explicit default return in this method, are we ok with `None` for this flag?

File: EquiTrack/partners/models.py [hunk 4]
```diff
@@ -469,68 +517,75 @@ def audit_done(cls, partner, assessment=None):
...
-                total += budget_record.unicef_cash
-            else:
-                q = q.values_list('unicef_cash', flat=True)
-                total = sum(q)
-
-        hact = json.loads(partner.hact_values) if isinstance(partner.hact_values, str) else partner.hact_values
-        hact["planned_cash_transfer"] = float(total)
-        partner.hact_values = hact
-        partner.save()
+    # @classmethod
```
Comment (reinbach) [inline EquiTrack/partners/models.py]: Could we remove this commented code completely and rely on git if we need to pull this code back in?

File: EquiTrack/vision/adapters/partner.py [hunk 5]
```diff
@@ -187,7 +187,8 @@ def _partner_save(processed, partner):
                     partner_org.vision_synced = True
                     saving = True
 
-                if partner_org.total_ct_cp is None or partner_org.total_ct_cy is None or \
+                if 1 < datetime.today().month <= 9 \
```
Comment (reinbach) [inline EquiTrack/vision/adapters/partner.py]: Can we add some comments explaining what exactly we're attempting to check here. Reading this code, is tricky

File: EquiTrack/partners/models.py [hunk 6]
```diff
@@ -469,68 +517,75 @@ def audit_done(cls, partner, assessment=None):
...
+        elif PartnerOrganization.CASH_TRANSFER_THRESHOLD_LEV1 < cash_transferred <= PartnerOrganization.CASH_TRANSFER_THRESHOLD_LEV2:
             spot_checks = 1
-        elif 100000.00 < cash_transferred <= 350000.00:
-            if self.rating in ['Low', 'Moderate']:
-                programme_visits = 1
-                spot_checks = 1
-            else:
+            programme_visits = 1
+        elif PartnerOrganization.CASH_TRANSFER_THRESHOLD_LEV2 < cash_transferred <= PartnerOrganization.CASH_TRANSFER_THRESHOLD_LEV3:
+            if self.rating in [_('High'), _('high'), _('Significant'), _('Non-Assessed'), _('Not Required'),
```
Comment (reinbach) [inline EquiTrack/partners/models.py]: How come we have a lowercase version of `High` in the list?
If it is required, maybe add a comment explaining why

File: EquiTrack/partners/models.py [hunk 7]
```diff
@@ -469,68 +517,75 @@ def audit_done(cls, partner, assessment=None):
...
                 spot_checks = 2
-        else:
-            if self.rating in ['Low', 'Moderate']:
-                programme_visits = 2
+            elif self.rating in [_('Low'), _('Moderate'), _('Medium')]:
+                programme_visits = 1
                 spot_checks = 1
-            else:
+        else:
+            if self.rating in [_('High'), _('high'), _('Significant'), _('Non-Assessed'),
```
Comment (reinbach) [inline EquiTrack/partners/models.py]: How come we have a lowercase version of `High` in the list?
If it is required, maybe add a comment explaining why

File: EquiTrack/partners/models.py [hunk 8]
```diff
@@ -591,13 +646,13 @@ def spot_checks(cls, partner, update_one=False):
...
 
-        partner.hact_values['spot_checks'] = sc
+        partner.hact_values['spot_checks']['completed']['total'] = sc
         partner.save()
 
-    @classmethod
-    def follow_up_flags(cls, partner, update_one=False):
-        partner.hact_values['follow_up_flags'] = 0
-        partner.save()
+    # @classmethod
```
Comment (reinbach) [inline EquiTrack/partners/models.py]: Could we remove this commented code completely and rely on git if we need to pull this code back in?

File: EquiTrack/hact/models.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import json
```
Comment (reinbach) [inline EquiTrack/hact/models.py]: Don't forget to add the __future__ imports on new files

File: EquiTrack/hact/serializers.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import json
```
Comment (reinbach) [inline EquiTrack/hact/serializers.py]: Don't forget to add the __future__ imports on new files

File: EquiTrack/hact/views.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
+from django_filters.rest_framework import DjangoFilterBackend
```
Comment (reinbach) [inline EquiTrack/hact/views.py]: Don't forget to add the __future__ imports on new files

File: EquiTrack/hact/management/commands/freeze_hact_data.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+from users.models import Country
+
+
+class Command(BaseCommand):
+    help = 'Freeze Hact Data for Current Year'
+
+    def add_arguments(self, parser):
+        parser.add_argument('--schema', dest='schema')
+
+    mapping_labels = {
```
Comment (reinbach) [inline EquiTrack/hact/management/commands/freeze_hact_data.py]: Maybe wrap text sections in `_(...)` to prep code for i18n

Comment (domdinicola) [reply EquiTrack/hact/management/commands/freeze_hact_data.py]: sure!

File: EquiTrack/hact/admin.py (hunk 1 above)
Comment (domdinicola) [reply EquiTrack/hact/admin.py]: sure!

File: EquiTrack/hact/models.py (hunk 9 above)
Comment (domdinicola) [reply EquiTrack/hact/models.py]: sure!

File: EquiTrack/hact/serializers.py (hunk 10 above)
Comment (domdinicola) [reply EquiTrack/hact/serializers.py]: sure!

File: EquiTrack/hact/views.py (hunk 11 above)
Comment (domdinicola) [reply EquiTrack/hact/views.py]: sure!

File: EquiTrack/partners/models.py (hunk 3 above)
Comment (domdinicola) [reply EquiTrack/partners/models.py]: let me ask the business, they told me last_assessment_date should always be there.. so I'm going to ask what to return if null

File: EquiTrack/partners/models.py (hunk 6 above)
Comment (domdinicola) [reply EquiTrack/partners/models.py]: I got some 'high'... most probably data cleaning is needed

File: EquiTrack/vision/adapters/partner.py (hunk 5 above)
Comment (domdinicola) [reply EquiTrack/vision/adapters/partner.py]: Sure!

File: EquiTrack/partners/models.py (hunk 6 above)
Comment (reinbach) [reply EquiTrack/partners/models.py]: Are we expecting translated versions of these texts as well? If not, we can change `self.rating` to `self.rating.lower()` and then have `['high', 'significant', ....]` and no need for different case versions of texts

File: EquiTrack/hact/views.py (hunk 2 above)
Comment (domdinicola) [reply EquiTrack/hact/views.py]: this is a temporary static json provided in order to help the front end guys to develop the charts. 
I'll write a comment in the code

File: EquiTrack/partners/models.py (hunk 3 above)
Comment (domdinicola) [reply EquiTrack/partners/models.py]: else return False

File: EquiTrack/partners/models.py (hunk 4 above)
Comment (domdinicola) [reply EquiTrack/partners/models.py]: sure!

File: EquiTrack/partners/models.py (hunk 8 above)
Comment (domdinicola) [reply EquiTrack/partners/models.py]: sure!

**Comment by domdinicola** [comment]
@caktus-philip I totally agree with PartnerOrganization.rating observation, the problem is that vision changes often the choices labels... and the business wants remapping as well
 
@reinbach I have to add tons of tests

**Comment by domdinicola** [comment]
@caktus-philip PartnerOrganization.rating as choice needs to wait prp-refactoring branch to be merged (it require a migration)

I want to add some more tests especially for:
1. get_current_quarter
2. HactHistorySerializer
3. HactHistoryAPIView
4. PartnerOrganization.expiring_assessment_flag
5. PartnerOrganization.approaching_threshold_flag
6. programmatic_visits by quarter

File: EquiTrack/hact/management/commands/freeze_hact_data.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    def freeze_data(self, hact_history):
+        partner = hact_history.partner
+        # partner values list needs to be in the
+        # desired order for export results
+        partner_values = [
+            ('Implementing Partner', partner.name),
+            ('Partner Type', partner.partner_type),
+            ('Shared', partner.shared_partner),
+            ('Shared IP', partner.shared_with),
+            ('TOTAL for current CP cycle', "{:.2f}".format(partner.total_ct_cp)),
```
[the remark this answers is not in the document]
Comment (reinbach) [reply EquiTrack/hact/management/commands/freeze_hact_data.py]: Was attempting to format the value, we can drop that or add better data handling

File: EquiTrack/hact/mappings.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
[the remark this answers is not in the document]
Comment (reinbach) [reply EquiTrack/hact/mappings.py]: Yes we should

File: EquiTrack/hact/management/commands/freeze_hact_data.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (reinbach) [reply EquiTrack/hact/management/commands/freeze_hact_data.py]: Ok

**Comment by domdinicola** [comment]
Great! :1st_place_medal: let's merge this!!!!!!!!!!!!

**Comment by domdinicola** [comment]
@reinbach thanks a lot

**Pull request merged**